packages/integration-react/src/api/ScmAuth.ts

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OAuthApi } from '@backstage/core-plugin-api';
+import {
+  ScmAuthApi,
+  ScmAuthTokenOptions,
+  ScmAuthTokenResponse,
+} from './ScmAuthApi';
+
+type ScopeMapping = {
+  default: string[];
+  repoWrite: string[];
+};
+
+class ScmAuthMux implements ScmAuthApi {
+  #providers = new Array<ScmAuth>();
+
+  constructor(providers: ScmAuth[]) {
+    this.#providers = providers;
+  }
+
+  async getCredentials(
+    options: ScmAuthTokenOptions,
+  ): Promise<ScmAuthTokenResponse> {
+    const url = new URL(options.url);
+    const provider = this.#providers.find(p => p.isUrlSupported(url));
+    if (!provider) {
+      throw new Error(
+        `No authentication provider available for access to '${options.url}'`,
+      );
+    }
+
+    return provider.getCredentials(options);
+  }
+}
+
+/**
+ * An implementation of the ScmAuthApi that merges together OAuthApi instances
+ * to form a single instance that can handles authentication for multiple providers.
+ */
+export class ScmAuth implements ScmAuthApi {
+  /**
+   * Creates a general purpose ScmAuth instance with a custom scope mapping.
+   */
+  static forAuthApi(
+    authApi: OAuthApi,
+    options: {
+      host: string;
+      scopeMapping: {
+        default: string[];
+        repoWrite: string[];
+      };
+    },
+  ): ScmAuth {
+    return new ScmAuth(authApi, options.host, options.scopeMapping);
+  }
+
+  /**
+   * Creates a new ScmAuth instance that handles authentication towards GitHub.
+   *
+   * The host option determines which URLs that are handled by this instance and defaults to `github.com`.
+   *
+   * The default scopes are:
+   *
+   * `repo read:org read:user`
+   *
+   * If the additional `repoWrite` permission is requested, these scopes are added:
+   *
+   * `gist`
+   */
+  static forGithub(
+    githubAuthApi: OAuthApi,
+    options?: {
+      host?: string;
+    },
+  ): ScmAuth {
+    const host = options?.host ?? 'github.com';
+    return new ScmAuth(githubAuthApi, host, {
+      default: ['repo', 'read:org', 'read:user'],
+      repoWrite: ['repo', 'read:org', 'read:user', 'gist'],
```

Comment:
I was a bit surprised here, I thought `repoWrite` would be additive and not need to repeat things from the default. I see in some cases below it's not strictly additive (e.g. Azure `vso.code` versus `vso.code_manage`) - or can it be in those cases as well?

Reply:
I kept them separate since they're not always strictly additive, but tbh it's prolly nicer to switch it around. Any auth provider should just ignore the extra scopes anyway